src/Parsers/ASTSelectQuery.cpp

```diff
@@ -113,7 +113,7 @@ void ASTSelectQuery::formatImpl(const FormatSettings & s, FormatState & state, F
     if (group_by_with_cube)
         s.ostr << (s.hilite ? hilite_keyword : "") << s.nl_or_ws << indent_str << (s.one_line ? "" : "    ") << "WITH CUBE" << (s.hilite ? hilite_none : "");
 
-    if (group_by_with_grouping_sets)
+    if (groupBy() && group_by_with_grouping_sets)
```

Comment:
That's completely unclear how `groupBy()` may be `nullptr` and `group_by_with_grouping_sets` is `true`. Maybe it's better to throw an exception here.

Reply:
Thanks. After looking at the code again, I agree that this situation should never occur.
Now throwing an exception.